pyvrp/tests/test_ProblemData.py

```diff
@@ -178,6 +178,34 @@ def test_problem_data_raises_when_incorrect_matrix_dimensions(matrix):
         ProblemData(clients, depots, vehicle_types, other_matrix, matrix)
 
 
+def test_problem_data_raises_matrix_diagonal_nonzero():
+    """
+    Tests that the ``ProblemData`` constructor raises a ``ValueError`` when
+    the distance or duration matrix has a non-zero value on the diagonal.
+    """
+    clients = [Client(x=0, y=0)]
+    depots = [Client(x=0, y=0)]
+    vehicle_types = [VehicleType(2, capacity=1)]
+
+    # distance matrix with non-zero value on the diagonal
+    distance_matrix = np.array([[0, 0], [0, 1]], dtype=int)
+    duration_matrix = np.zeros((2, 2), dtype=int)
+
+    with assert_raises(ValueError):
+        ProblemData(
+            clients, depots, vehicle_types, distance_matrix, duration_matrix
+        )
+
+    # duration matrix with non-zero value on the diagonal
+    distance_matrix = np.zeros((2, 2), dtype=int)
+    duration_matrix = np.array([[0, 0], [0, 1]], dtype=int)
+
+    with assert_raises(ValueError):
+        ProblemData(
+            clients, depots, vehicle_types, distance_matrix, duration_matrix
+        )
```

Comment:
This can be halved using `pytest.mark.parametrize` to pass in the matrices. That'd make the separation between the data causing the issue and the set-up to get to where we can test it a bit clearer.